### DISCUSSION TITLE
Forming a project maintainer team

### DISCUSSION THREAD

**Discussion by awvwgk** [fortran-lang/fpm#542] [General]
We have been discussing to form a maintainer team for fpm in a couple of discussions already.
I linked the relevant ones here:

- Call for reviewers / maintainers at stdlib: https://fortran-lang.discourse.group/t/1506
- Fortran monthly call July 2021: https://fortran-lang.discourse.group/t/1510
- (upcoming) Fortran monthly call August 2021: https://fortran-lang.discourse.group/t/1628
- Request for a decision making progress: https://github.com/fortran-lang/fpm/issues/526
- Future roadmap: https://github.com/fortran-lang/fpm/discussions/338

To move this forward I suggest to take the following steps

1. anyone interested in taking up maintainer responsibility for fpm can volunteer here
2. we create a (public) maintainer team for the respective projects in GitHub,
    this allows tagging `@fortran-lang/fpm` or similar in issues or PRs
    (I think there is already a @fortran-lang/fpm-developers team around)
3. we review the repository for:
    - updating the contribution guidelines, README and docs
    - adding a (near future) roadmap to the project wiki

I have the feeling that with fpm we already have a team of de-facto maintainers here.
The thing that would essentially change for us, is how we steer the development of fpm and set goals for the project.

Let me know what you think.

**Comment by everythingfunctional** [comment]
Sounds good to me, and I will of course be happy to join the team.

**Comment by LKedward** [comment]
Thanks Sebastian! This all looks good to me and, as previously communicated, I'm happy to volunteer for this.

**Comment by certik** [comment]
I don't have time currently to do day to day work on `fpm`, however I want to be involved in its direction, because the direction and design of `fpm` is crucial for success of the Fortran package ecosystem. So I want to be part of all design discussions and decisions.

**Comment by milancurcic** [comment]
Thanks, @awvwgk!

And to be clear, anybody (maintainer and non-maintainer alike) can be part of all design discussions and decisions.

**Comment by awvwgk** [comment]
>  The thing that would essentially change for us, is how we steer the development of fpm and set goals for the project.

Sorry for being unclear here. I was more thinking about setting priorities for features to develop or defining milestones, like being able to build class of projects, which would fall under a near future roadmap. I think this is an essential freedom for project maintainers to have. 

Of course, all design discussions and decisions will happen publicly and are open for everybody to participate in as before.

**Comment by certik** [comment]
@awvwgk, @milancurcic in your mind, what exactly is the role of the maintainers?

It looks like we all agree that anybody can be involved in discussions and decisions. We probably would agree that anybody can be involved with sending PRs and reviewing PRs. In terms of contributors, say over the last 8 months (from January 1, 2021), the [top 4 are](https://github.com/fortran-lang/fpm/graphs/contributors?from=2021-01-01&to=2021-08-16&type=c): @LKedward, @urbanjost, @awvwgk and @brocolis. The order stays the same over more recent periods. 

In terms of decision making, it seems there wasn't a problem so far for more day to day activities. For big issues like `.f` for free form requires discussion with the wide community anyway and should be made beyond any small number of current maintainers anyway.

It seems the main need is to get reviewers for day to day activities, and also somebody who is willing to have a vision and lead the development. Somebody (team) that we trust to be able to keep the ball rolling and ensuring the community would agree with the major decisions. Small things can always be fixed.

**Reply by certik** [reply]
Pulling the thread a little more:

> Sorry for being unclear here. I was more thinking about setting priorities for features to develop or defining milestones, like being able to build class of projects, which would fall under a near future roadmap. I think this is an essential freedom for project maintainers to have.

It seems those should be the people who actually contribute code, so the 4 above (or more) that I mentioned. One idea from Linus Torvalds is that the person who does the work decides about how it gets done. Obviously, we must trust each other that we don't screw things up and that we will reach for the community for big design questions. But in terms of priorities, at the end of the day, what gets done depends on who submits the PRs. The roadmap is something that helps to motivate. And hopefully people who work on the project share the vision. :)

**Reply by awvwgk** [reply]
Please, let's not start counting commits here, there is no point. Contributions to fpm were never limited to code contributions only. The issue tracker and this discussion board as well as the discourse were always an integral part of the development of fpm with many great contributions to this project, which are not captured in GitHub's contributor metric at all, because it is limited to commits and LOC changes only.

**Reply by certik** [reply]
Indeed, every metric is imperfect. We can count issue tracker contributions, you can use this tool: https://fortran-lang.org/community/, from January that would be @awvwgk, @LKedward, @urbanjost and @milancurcic.

A pattern emerges. Given this, I would suggest @urbanjost to join the maintainers team also.

We should be inclusive in setting up these teams, and one way to ensure that frequent contributors are included is by finding them with such metrics, as imperfect as they individually are. (Of course I agree and have always advocated for contributions to being defined broadly.)

**Reply by awvwgk** [reply]
Yes, of course, John @urbanjost has been a core contributor and developer for fpm. I hope to have him on board with this as well.

We should take some time here and make sure everybody gets a chance to comment in this thread. There is no need for hurry.

**Comment by awvwgk** [comment]
> what exactly is the role of the maintainers?

This is still open for discussion in way.

A few points I find particular important are
- help new contributors with their first contributions and get acquainted with the project
- review code contributions or find / assign reviewers
- prioritize on projects and coordinate larger code base changes

**Reply by certik** [reply]
Yes, I think I can agree on that.

**Comment by awvwgk** [comment]
I've updated the @fortran-lang/fpm team, this team can now be pinged and requested reviews from.
